ERCS/erc-simulate-calls.md

```diff
@@ -0,0 +1,227 @@
+
+## Motivation
+
+Applications are reliant on JSON-RPC communication to Wallets in order to execute actions. A seemingly core functionality for "offline" Applications (Servers, etc) is the ability to simulate if an action will succeed or fail and/or estimate the total fee and/or calculate balance changes, prior to signing and broadcasting to the network.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### `wallet_simulateCalls`
+
+Instructs a Wallet to simulate execution of a set of calls, and return metadata such as: the gas used, logs, status, and failure reason (if applicable).
+
+#### Request
+
+- Consumers MUST provide exact parameters as per [ERC-5792 `wallet_sendCalls`](https://eips.ethereum.org/EIPS/eip-5792#wallet_sendcalls).
+- Consumers MAY OPTIONALLY provide [ERC-7836](https://eips.ethereum.org/ERCs/eip-7836) `context` and `signature` values from a "prepared" call.
+  - `context` is the result of `wallet_prepareCalls`.
+  - `signature` is the result of signing a [ERC-7836](https://eips.ethereum.org/ERCs/eip-7836) `digest` with the sender's private key.
+
+```ts 
+type Request = {
+  method: 'wallet_simulateCalls',
+  params: [{
+    // Calls to simulate.
+    calls: {
+      to: `0x${string}`,
+      data?: `0x${string}`,
+      value?: `0x${string}`,
+      capabilities?: Record<string, any>;
+    }[],
+    // Target chain ID to simulate calls on.
+    chainId: `0x${string}`,
+    // ERC-5792 Capabilities request.
+    capabilities?: Record<string, any>;
+    // ERC-7836 context.
+    context?: unknown;
+    // Sender address.
+    from?: `0x${string}`;
+    // ERC-7836 signature.
+    signature?: {
+      publicKey: `0x${string}`,
+      type: string,
+      value: `0x${string}`
+    };
+    // Version.
+    version: string;
+  }]
+}
+```
+
+#### Response
+
+- The `status` code complies with the [ERC-5792 Status Codes](https://github.com/ethereum/EIPs/blob/2dcee4d0e2fc1cea488c12ba88e9a93d5925043b/EIPS/eip-5792.md#status-codes-for-status-field).
+- A JSON-RPC `error` object is returned if the simulation fails.
+  - The `error` object MAY be present if the `status` code is `4xx | 5xx | 6xx`.
+  - `error.message` MUST be a short human-readable message of the error.
+  - `error.details` MAY contain low-level details of the error.
+  - `error.data` MUST be an hex-encoded error (selector) if the `status` code is an onchain failure (`5xx | 6xx`).
+
+```ts
+type Response = {
+  // Chain ID the calls were simulated on.
+  chainId: `0x${string}`;
+  // Estimated total gas used by the calls.
+  gasUsed: `0x${string}`;
+  // Logs emitted by the calls.
+  logs: {
+    address: `0x${string}`;
+    data: `0x${string}`;
+    topics: `0x${string}`[];
+  }[];
```

Comment:
Call traces are also useful for surfacing internal contract calls and internal eth transfers, if the simulator supports outputting them. We could mark that as an optional field in the response (or make this a capability?)